src/txn/txn_ckpt.c

```diff
@@ -627,6 +629,22 @@ __checkpoint_prepare(WT_SESSION_IMPL *session, bool *trackingp, const char *cfg[
 
     __wt_writeunlock(session, &txn_global->rwlock);
 
+    /*
+     * Open an internal session to allocate a transaction id that is used for removing the history
+     * store entries due to a prepared rollback occurs in parallel to the checkpoint. Make sure that
+     * this transaction id is published before the checkpoint acquires its snapshot.
+     */
+    WT_RET(__wt_open_internal_session(conn, "checkpoint internal", true, session->flags,
+      session->lock_flags, &ckpt_internal_session));
+    WT_ERR(__wt_txn_begin(ckpt_internal_session, NULL));
+    WT_ERR(__wt_txn_id_check(ckpt_internal_session));
+    txn_global->checkpoint_invisible_txn_id = ckpt_internal_session->txn->id;
+
+    /* Add a one second wait to simulate invisible transaction id race with prepared rollback. */
+    tsp.tv_sec = 1;
+    tsp.tv_nsec = 0;
+    __checkpoint_timing_stress(session, WT_TIMING_STRESS_CHECKPOINT_INVISIBLE_TXNID_DELAY, &tsp);
+
```

Comment:
Can we add an assert that the invisible traction id is not visible to checkpoint snapshot?

Reply:
Added an assert to confirm the reserved transaction id is not visible according to the checkpoint snapshot.